DWELL: Pay your rent with a yield-bearing stablecoin

rohan-patra opened the pull request:
## Summary

Added new token standard implementation and Foundry project setup:

1. **SERC20 Token Standard**
   - Introduced SIERC20 interface
   - Implemented privacy-preserving ERC20 with shielded balances/transfers
   - Added USDY yield-bearing stablecoin using shares-based accounting

2. **Project Infrastructure**
   - Set up Foundry project with:
     - Foundry.toml config
     - .env.example with local RPC
     - Comprehensive .gitignore
   - Added test suite and deployment scripts
   - Included forge-std library as dependency

3. **Documentation**
   - Updated root README with token project details
   - Added dedicated tokens/README.md explaining:
     - Privacy/yield challenges
     - Architecture (SERC20/USDY)
     - License (AGPL-3.0)

raviriley reviewed (approved):
![image](https://github.com/user-attachments/assets/1738445f-6dcb-41b7-8ac3-d801931871e8)
LGTM

lyronctk commented:
> ![image](https://private-user-images.githubusercontent.com/8132955/411198124-1738445f-6dcb-41b7-8ac3-d801931871e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkwMzYyMjEsIm5iZiI6MTczOTAzNTkyMSwicGF0aCI6Ii84MTMyOTU1LzQxMTE5ODEyNC0xNzM4NDQ1Zi02ZGNiLTQxYjctOGFjMy1kODAxOTMxODcxZTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDhUMTczMjAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTNiYmRjNDM1NmY5M2FiNzdlMTU3MTY3NDFmYmExZWYxZjliNTVkNTBkMDExODc5YzZhOWZmNDU4MGQwNTcxZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.szBn3RDlJoTf2X3_0umOwTVrzhpiXlsKfTW5HJejTTI) LGTM

LOL

ssolit commented on tokens/src/SRC20.sol:
```diff
@@ -0,0 +1,405 @@
+     * @dev Atomically decreases the allowance granted to a shielded `spender` by a shielded `subtractedValue`.
+     *
+     * This is an alternative to {approve} that can be used as a mitigation for
+     * problems described in {SIRC20-approve}.
```
"problems described in {SIRC20-approve}" Where is this? I'm having a bit of trouble finding it

rohan-patra replied:
Sorry, we seem to have removed the warning from the original ERC20 contract, I just added it back.

``` 
WARNING: Changing an allowance with this method can have security implications. When changing an approved
allowance to a specific value, a race condition may occur if another transaction is submitted before
the original allowance change is mined. To safely adjust allowances, use {increaseAllowance} and
{decreaseAllowance} which provide atomic operations protected against such race conditions.
```

ssolit commented:
Would propose the following two revisions:

1. revert in private view functions instead of returning 0
This would help dApps know whether the value is really zero or if something went wrong. It would be unfortunate for a dApp to believe something is really zero when it's not and use it for business logic. The contract could return a custom error, perhaps named UnauthorizedView?

How dApps could interface:
a) Caller contracts can use .call() to error handle if they want
b) The SRC20 could also provide a safeBalanceOf function that returns (bool success, uint256 balance)


2. For ISRC20, instead of providing events, have emitEvent functions that must be overridden
For the default SRC20, these should be no-ops because we assume users want complete privacy from indexers. However, these functions can be overridden with custom behavior, which leaves the door open to future improvements (i.e. encrypted events)
Requires replacing all emits in the SRC20 with calls to emitApprovalEvent, etc

rohan-patra commented:
> Would propose the following two revisions:
> 
> 1. revert in private view functions instead of returning 0
>    This would help dApps know whether the value is really zero or if something went wrong. It would be unfortunate for a dApp to believe something is really zero when it's not and use it for business logic. The contract could return a custom error, perhaps named UnauthorizedView?
> 
> How dApps could interface: a) Caller contracts can use .call() to error handle if they want b) The SRC20 could also provide a safeBalanceOf function that returns (bool success, uint256 balance)
> 
> 2. For ISRC20, instead of providing events, have emitEvent functions that must be overridden
>    For the default SRC20, these should be no-ops because we assume users want complete privacy from indexers. However, these functions can be overridden with custom behavior, which leaves the door open to future improvements (i.e. encrypted events)
>    Requires replacing all emits in the SRC20 with calls to emitApprovalEvent, etc

I modified the original function to return the error, and added a new function that returns 0 on unauthorized attempts and also returns a boolean to indicate that the call was unsuccessful

Also updated the tests to expect the appropriate reverts

ssolit commented on rent/src/USDY.sol:
```diff
@@ -0,0 +1,455 @@
+    }
+
+    /**
+     * @dev Override _update to handle shielded shares accounting
```
Maybe instead of "handle shielded shares accounting", say "account for shielded shares rather than raw token balances"

ssolit commented on rent/src/USDY.sol:
```diff
@@ -0,0 +1,455 @@
+
+    /**
+     * @notice Transfers a specified number of tokens from the caller's address to the recipient.
+     * @dev Uses _update for share-based accounting while maintaining token-based interface
```
Took me a minute to understand the @dev. maybe say "Uses new _update for share-based accounting while maintaining token-based amount parameter"

ssolit replied:
ditto for transferFrom

ssolit commented on rent/src/USDY.sol:
```diff
@@ -0,0 +1,455 @@
+    }
+
+    /**
+     * @notice This is the exact same implementation as the one in SRC20.sol
```
If it's the exact same, would you want to just leave it out completely? alternatively, if you want to be explicit, you could call super.allowance(). Less code duplication this way

ssolit commented on rent/src/USDY.sol:
```diff
@@ -0,0 +1,455 @@
+    }
+
+    /**
+     * @notice Sets `amount` as the allowance of `spender` over the caller's tokens
```
would appreciate an @dev comment explaining why this is different from SRC20. prevents invalid spender?

ssolit replied:
ditto for increaseAllowance, decreaseAllowance

ssolit commented on rent/src/USDY.sol:
```diff
@@ -0,0 +1,455 @@
+     */
+    function approve(saddress spender, suint256 amount) public override whenNotPaused returns (bool) {
+        address owner = _msgSender();
+        if (spender == saddress(address(0))) {
```
This code looks similar to super._approve(). I see you have the whenNotPaused modifier, but consider calling super._approve() in this function

ssolit reviewed (commented):
comments for USDY

ssolit commented on rent/src/SRC20.sol:
```diff
@@ -0,0 +1,449 @@
+ * @dev Implementation of the {ISRC20} interface with privacy protections using shielded types.
+ * Public view functions that would leak privacy are implemented as no-ops while maintaining interface compatibility.
+ * Total supply remains public while individual balances and transfers are private.
+ * Currently, this implementation is fully compliant with the ERC-20 standard, meaning that transfer recipients are NOT shielded/private.
```
We can remove this comment after the events refactor?

ssolit commented on rent/src/SRC20.sol:
```diff
@@ -0,0 +1,449 @@
+            }
+        }
+
+        emitTransfer(address(from), address(to), 0); // Replace _emitTransferEvent
```
I would input the real value instead of 0 now, since the modifications to the event are done in emitTransfer

ssolit commented on rent/src/SRC20.sol:
```diff
@@ -0,0 +1,449 @@
+     * (or `to`) is the zero address. All customizations to transfers, mints, and burns should be done by overriding
+     * this function.
+     *
+     * Emits a {Transfer} event.
```
update to say calls emitTransferEvent

ssolit replied:
Generally I think a lot of the docs need to be updated

ssolit commented on rent/src/SRC20.sol:
```diff
@@ -0,0 +1,449 @@
+     * This internal function is equivalent to `approve`, and can be used to
+     * e.g. set automatic allowances for certain subsystems, etc.
+     *
+     * Emits an {Approval} event.
```
say calls emitApprovalEvent, which in this implementation is a noop. 

also, we probably don't need two _approval functions given this behavior

ssolit commented on rent/script/SRC20.s.sol:
remember to remove scripts

ssolit commented on rent/src/SRC20.sol:
```diff
@@ -0,0 +1,421 @@
+     * @dev Creates a shielded `value` amount of tokens and assigns them to a shielded `account`.
+     * Relies on the `_update` mechanism.
+     *
+     * Emits a {Transfer} event with zero values to maintain privacy.
```
No longer emits zero values

ssolit replied:
ditto for _transfer

ssolit commented on rent/src/USDY.sol:
```diff
@@ -0,0 +1,432 @@
+    function approve(saddress spender, suint256 amount) public override whenNotPaused returns (bool) {
+        address owner = _msgSender();
+        if (owner == address(0)) revert ERC20InvalidSpender(address(0));
+        emit Approval(owner, address(spender), uint256(amount));
```
will this emit an event twice since super.approve might emit an event?
if not, should this me a call to the emitEvent function?

ditto for increaseAllowance and decreaseAllowance?

ssolit commented on rent/src/ISRC20.sol:
```diff
@@ -0,0 +1,83 @@
+     *
+     * Returns a boolean value indicating whether the operation succeeded.
+     *
+     * Emits a {Transfer} event with zero values to maintain privacy.
```
I still see "Emits a {Transfer} event with zero values to maintain privacy."
